src/libraries/System.Private.CoreLib/src/System/DateTime.cs

```diff
@@ -309,79 +284,52 @@ public DateTime(int year, int month, int day, int hour, int minute, int second,
         {
             if (calendar == null)
                 throw new ArgumentNullException(nameof(calendar));
-            if (millisecond < 0 || millisecond >= MillisPerSecond)
-            {
-                throw new ArgumentOutOfRangeException(nameof(millisecond), SR.Format(SR.ArgumentOutOfRange_Range, 0, MillisPerSecond - 1));
-            }
 
-            int originalSecond = second;
-            if (second == 60 && s_systemSupportsLeapSeconds)
+            if (second != 60 || !s_systemSupportsLeapSeconds)
             {
-                // Reset the second value now and then we'll validate it later when we get the final Gregorian date.
-                second = 59;
+                _dateData = calendar.ToDateTime(year, month, day, hour, minute, second, millisecond).UTicks;
             }
-
-            long ticks = calendar.ToDateTime(year, month, day, hour, minute, second, 0).Ticks;
-            ticks += millisecond * TicksPerMillisecond;
-            if (ticks < MinTicks || ticks > MaxTicks)
-                throw new ArgumentException(SR.Arg_DateTimeRange);
-            _dateData = (ulong)ticks;
-
-            if (originalSecond == 60)
+            else
             {
-                DateTime dt = new DateTime(_dateData);
-                if (!IsValidTimeWithLeapSeconds(dt.Year, dt.Month, dt.Day, dt.Hour, dt.Minute, 60, DateTimeKind.Unspecified))
-                {
-                    throw new ArgumentOutOfRangeException(null, SR.ArgumentOutOfRange_BadHourMinuteSecond);
-                }
+                // if we have a leap second, then we adjust it to 59 so that DateTime will consider it the last in the specified minute.
+                this = new(year, month, day, hour, minute, 59, 999, calendar);
```

Comment:
ditto. pass the input milliseconds. 

I'll stop commenting on that as we need to fix that in all instances. We need only to change the milliseconds and extra ticks in case of `UtcNow` only.